bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -157,6 +161,53 @@ class BusinessPartnerService(
             ?.let { throw BpdmMissingStageException(it, StageType.Input) }
     }
 
+    /**
+     * Filters all [entities] by looking for existing business partner data in the given [stage]
+     *
+     * Filters incoming Business Partner for changes against the same persisted record in the Database
+     * If the data is the same, the consequent Business Partner will not have a new changelog and the golden record
+     * process will not start. If Business Partner has the same data, but linked sharing state has an error state,
+     * Business Partner will start the process again.
+     */
+    private fun filterUpdateCandidates(entities: List<BusinessPartner>, stage: StageType): List<BusinessPartner> {
+        val externalIds = entities.map { it.externalId }
+        val persistedBusinessPartner = businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds)
+        val sharingStates = sharingStateRepository.findByExternalIdInAndBusinessPartnerType(externalIds, BusinessPartnerType.GENERIC)
```

Comment:
You could change both lists to maps by adding `.associateBy({ it.external })`. This could make lookup simpler and more efficient.